internal/pbinfo/comments.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pbinfo
+
+import (
+	"fmt"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+)
+
+const (
+	// SourcePathTopItemLength the length of a top-level type path
+	SourcePathTopItemLength = 2
+	// SourcePathTopItemTypeIndex the index of the source code
+	// item path type
+	SourcePathTopItemTypeIndex = 0
+	// SourcePathTopItemIndex the index of a source code path
+	// referncing a specific Message
+	SourcePathTopItemIndex = 1
+	// SourcePathSubItemLength the length of a source code path
+	// referencing a sub-item of a Message or Service
+	SourcePathSubItemLength = 4
+	// SourcePathSubItemTypeIndex the index of the sub-item type
+	// in a source code path
+	SourcePathSubItemTypeIndex = 2
+	// SourcePathSubItemIndex the index of a source code path
+	// referencing a specific Field
+	SourcePathSubItemIndex = 3
+	// SourcePathServiceTypeValue the value that indicates a
+	// a source code path points to a Service
+	SourcePathServiceTypeValue = 6
+	// SourcePathMessageTypeValue the value that indicates a
+	// source code path points to a Message
+	SourcePathMessageTypeValue = 4
+	// SourcePathMethodTypeValue the value the indicates a
+	// source code path points to a Method
+	SourcePathMethodTypeValue = 2
+	// SourcePathFieldTypeValue represents the value associated
+	// with the Field type in a source code path
+	SourcePathFieldTypeValue = 2
+	// SourcePathExtensionTypeValue represents the value associated
+	// with the Extension type in a source code path
+	SourcePathExtensionTypeValue = 7
+	// SourcePathEnumTypeValue represents the value associated
+	// with the Enum type in a source code path
+	SourcePathEnumTypeValue = 5
+	// SourcePathEnumValueTypeValue represents the value associated
+	// with the Enum Value type in a source code path
+	SourcePathEnumValueTypeValue = 2
+)
+
+func addComments(comments map[string]string, f *descriptor.FileDescriptorProto, loc *descriptor.SourceCodeInfo_Location) {
```

Comment:
If you don't mind, I think we should move this out of `pbinfo` for now. I think the question of how we should represent comments are still not well-understood.

I believe it might be better to use `map[proto.Message]string` like https://github.com/googleapis/gapic-generator-go/blob/eaeab394a8a4df04f7260ed36245c17de8041699/internal/gengapic/gengapic.go#L145, so that we can look up directly. Otherwise, we have to make sure keys for two types cannot conflict. I think this implementation will run into trouble if we have
```
message A {
  message X {}
}
message B {
  message X{}
}
```

The two X's, I think, will end up with the same key.